File: Tests/LedgelingsCoreTests/HideoutTests.swift
import Testing
@testable import LedgelingsCore

/// The little house: appears, swallows everyone, packs itself away, and lets them out later.
@Suite struct HideoutTests {
    @Test func theWholeCycleFromHideToWelcomeBack() {
        var h = Hideout()
        #expect(!h.isActive && h.scale(at: 0) == 0)
        h.hide(count: 3, at: 100, for: 60)
        #expect(h.isActive && h.phase == .appearing)
        #expect(abs(h.scale(at: 100.2) - 0.5) < 1e-9)
        #expect(h.update(at: 100.4).isEmpty && h.phase == .gathering)
        #expect(h.scale(at: 101) == 1)
        h.entered(0, at: 101); h.entered(1, at: 102)
        #expect(h.phase == .gathering && h.isInside(1) && !h.isInside(2))
        h.entered(2, at: 103)
        #expect(h.phase == .shrinking)
        #expect(abs(h.scale(at: 103.25) - 0.5) < 1e-9)
        _ = h.update(at: 103.5)
        #expect(h.phase == .hidden && h.scale(at: 120) == 0)
        #expect(h.remaining(at: 120) == 40)
        _ = h.update(at: 159)
        #expect(h.phase == .hidden)
        _ = h.update(at: 160)
        #expect(h.phase == .growing)
        _ = h.update(at: 160.4)
        #expect(h.phase == .releasing)
        var out: [Int] = []
        for t in stride(from: 160.4, through: 163.0, by: 0.1) {
            for event in h.update(at: t) { if case .letOut(let i) = event { out.append(i) } }
        }
        #expect(out == [0, 1, 2])
        #expect(h.phase == .vanishing || h.phase == .away)
        _ = h.update(at: 164)
        #expect(h.phase == .away && !h.isActive && h.scale(at: 164) == 0)
    }

    @Test func stragglersAreForcedInsideWhenTheGatheringTakesTooLong() {
        var h = Hideout()
        h.hide(count: 2, at: 0, for: 600)
        _ = h.update(at: 0.4)
        h.entered(0, at: 1)
        let events = h.update(at: 0.4 + h.gatherCap)
        #expect(events.contains(.forceInside([1])))
        h.entered(1, at: 26)
        #expect(h.phase == .shrinking)
    }

    @Test func bringingThemBackEarlyOpensTheHouseNow() {
        var h = Hideout()
        h.hide(count: 1, at: 0, for: 3600)
        _ = h.update(at: 0.4); h.entered(0, at: 1); _ = h.update(at: 1.5)
        #expect(h.phase == .hidden)
        h.recall(at: 10)
        #expect(h.phase == .growing)
        #expect(h.remaining(at: 10) == 0)
    }

    @Test func recallingDuringTheGatheringJustLetsThemStay() {
        var h = Hideout()
        h.hide(count: 2, at: 0, for: 3600)
        _ = h.update(at: 0.4); h.entered(0, at: 1)
        h.recall(at: 2)
        #expect(h.phase == .releasing)
        let events = h.update(at: 2)
        #expect(events.contains(.letOut(0)))
    }

    @Test func hidingAgainWhileActiveIsIgnored() {
        var h = Hideout()
        h.hide(count: 2, at: 0, for: 60)
        h.hide(count: 5, at: 1, for: 5)
        #expect(h.count == 2 && h.remaining(at: 1) == 59 + 0.4 + 0.5 || h.remaining(at: 1) == 59)
    }
}

File: win/LedgelingsCore.Tests/HideoutTests.cs
namespace Ledgelings.Core.Tests;

/// <summary>The little house: appears, swallows everyone, packs itself away, and lets them out later.</summary>
public class HideoutTests
{
    [Fact]
    public void TheWholeCycleFromHideToWelcomeBack()
    {
        var h = new Hideout();
        Assert.True(!h.IsActive && h.Scale(0) == 0);
        h.Hide(3, 100, 60);
        Assert.True(h.IsActive && h.CurrentPhase == Hideout.Phase.Appearing);
        Assert.True(Math.Abs(h.Scale(100.2) - 0.5) < 1e-9);
        Assert.True(h.Update(100.4).Count == 0 && h.CurrentPhase == Hideout.Phase.Gathering);
        Assert.Equal(1, h.Scale(101));
        h.Entered(0, 101); h.Entered(1, 102);
        Assert.True(h.CurrentPhase == Hideout.Phase.Gathering && h.IsInside(1) && !h.IsInside(2));
        h.Entered(2, 103);
        Assert.Equal(Hideout.Phase.Shrinking, h.CurrentPhase);
        Assert.True(Math.Abs(h.Scale(103.25) - 0.5) < 1e-9);
        _ = h.Update(103.5);
        Assert.True(h.CurrentPhase == Hideout.Phase.Hidden && h.Scale(120) == 0);
        Assert.Equal(40, h.Remaining(120));
        _ = h.Update(159);
        Assert.Equal(Hideout.Phase.Hidden, h.CurrentPhase);
        _ = h.Update(160);
        Assert.Equal(Hideout.Phase.Growing, h.CurrentPhase);
        _ = h.Update(160.4);
        Assert.Equal(Hideout.Phase.Releasing, h.CurrentPhase);
        var letOut = new List<int>();
        // stride(from: 160.4, through: 163.0, by: 0.1)
        for (int k = 0; 160.4 + k * 0.1 <= 163.0 + 1e-9; k++)
        {
            var t = 160.4 + k * 0.1;
            foreach (var e in h.Update(t)) if (e is Hideout.Event.LetOut(var i)) letOut.Add(i);
        }
        Assert.Equal(new[] { 0, 1, 2 }, letOut);
        Assert.True(h.CurrentPhase == Hideout.Phase.Vanishing || h.CurrentPhase == Hideout.Phase.Away);
        _ = h.Update(164);
        Assert.True(h.CurrentPhase == Hideout.Phase.Away && !h.IsActive && h.Scale(164) == 0);
    }

    [Fact]
    public void StragglersAreForcedInsideWhenTheGatheringTakesTooLong()
    {
        var h = new Hideout();
        h.Hide(2, 0, 600);
        _ = h.Update(0.4);
        h.Entered(0, 1);
        var events = h.Update(0.4 + h.GatherCap);
        Assert.Contains(new Hideout.Event.ForceInside(new[] { 1 }), events);
        h.Entered(1, 26);
        Assert.Equal(Hideout.Phase.Shrinking, h.CurrentPhase);
    }

    [Fact]
    public void BringingThemBackEarlyOpensTheHouseNow()
    {
        var h = new Hideout();
        h.Hide(1, 0, 3600);
        _ = h.Update(0.4); h.Entered(0, 1); _ = h.Update(1.5);
        Assert.Equal(Hideout.Phase.Hidden, h.CurrentPhase);
        h.Recall(10);
        Assert.Equal(Hideout.Phase.Growing, h.CurrentPhase);
        Assert.Equal(0, h.Remaining(10));
    }

    [Fact]
    public void RecallingDuringTheGatheringJustLetsThemStay()
    {
        var h = new Hideout();
        h.Hide(2, 0, 3600);
        _ = h.Update(0.4); h.Entered(0, 1);
        h.Recall(2);
        Assert.Equal(Hideout.Phase.Releasing, h.CurrentPhase);
        var events = h.Update(2);
        Assert.Contains(new Hideout.Event.LetOut(0), events);
    }

    [Fact]
    public void HidingAgainWhileActiveIsIgnored()
    {
        var h = new Hideout();
        h.Hide(2, 0, 60);
        h.Hide(5, 1, 5);
        Assert.True(h.Count == 2 && h.Remaining(1) == 59 + 0.4 + 0.5 || h.Remaining(1) == 59);
    }
}
